versions/draft-3/SPEC.md

```diff
@@ -1515,6 +1522,130 @@ workflow wf {
 
 In this example, the fully-qualified names that would be exposed as workflow outputs would be `wf.task1.results`, `wf.altname.value`.
 
+## Struct Definition
+A struct is a C-like construct which enables the user to create new compound types that consist of previously existing types. Structs
+can then be used within a `Task` or `Workflow` definition as a declaration in place of any other normal types. The struct takes the place of the
+`Object` type in many circumstances and enables proper typing of its members.
+
+Structs are defined using the `struct` keyword and have a single section consiting of typed declarationns. They are evaluated first prior
+to workflows and tasks.
+```wdl
+struct name { ... }
+```
+
+### Struct Declarations
+The only contents of struct are a set of declarations. Declarations can be any primitive or compound type, as well as other structs, and are defined
+the same way as they are in any other section. The one caveat to this is that declarations within a struct do not allow an expression statement after
+the initial member declaration.
+
+for example the following is a valid struct definition
+```wdl
+struct name {
+    String myString
+    Int myInt
+}
+```
+Whereas the following is invalid
+
+```wdl
+struct invalid {
+    String myString = "Cannot do this"
+    Int myInt
+}
+```
+
+You can also use compound types within a struct to easily encapsulate them within a single object. For example
+```wdl
+struct name {
+    Array[Array[File]] myFiles
+    Map[String,Pair[String,File]] myComplexType
+    String cohortName
+}
+```
+#### Optional and non Empty Struct Values
+Struct declarations can be optional or non-empty (if they are an array type). 
+
+```wdl
+struct name {
+    Arrray[File]+ myFiles
+    Boolean? myBoolean
+}
+```
+
+### Using a Struct
+When using a struct in the declaration section of either a `workflow` or a `task` or `output` section you define them in the same way you would define any other type.
+
+For example, if I have a struct like the following:
+```wdl
+struct MyStruct {
+    String myName
+    Int myAge
+}
+```
+
+then usage of the struct in a workflow would look like the following:
+
+```wdl
+
+task myTask {
+    MyStruct a
+    command {
+        echo "hello my name is ${a.myName} and I am ${a.myAge} years old"
+    }
+}
+
+workflow myWorkflow {
+    MyStruct a
+    call myTask {
+        input:
+            a = a
+    }
+
+}
+```
+
+#### Struct Assignment from Object Literal
+Structs can be assigned using an object literal. When Writing the object, all entries must conform or be coercible into the underlying type they are being assigned to
```

Comment:
is that the only way to assign to a struct? Could one assign to it "slowly"? 
For example:

```wdl 
MyStruct a
a.myName="John"
a.myAge=30
```

Reply:
I would say you cannot assign slowly, at least not at the moment

Reply:
I don't like this. WDL values are immutable and I think they should stay that way.

Reply:
OK. in that case, please also include an example of building a complex struct using other (previously defined) values:

```wdl
Array[File] someFiles=["file1.txt","file2.txt"]
Int someAge=30

struct MyStruct{
Array[File] files
Int age
}

MyStruct a={"files"=someFiles "age"=someAge}
```

Also given that the key/value pair has quotes on the key, it is possible to parametrize the assignments:

```wdl
Array[File] someFiles=["file1.txt","file2.txt"]
Int someAge=30
String key1="files"
String key2="age"
struct MyStruct{
Array[File] files
Int age
}

MyStruct a={key1=someFiles key2=someAge}
```
? if not, I'd advocate for removing the quotes from the keys.

Reply:
I like this idea of removing quotes from the keys, not least because it makes sure keys are valid identifiers rather than arbitrary strings.

Reply:
@geoffjentry I would be interested to know what is currently allowable syntax with Object Literal Notation. Does it need to be JSON like? Is it possible to remove the quotes from keys under the current spec

Reply:
FWIW I also prefer immutability.

@patmagee That's a good question. I realized yesterday that object isn't actually *defined* anywhere, period. It also doesn't help that AFAIK Broad folks aren't using object so it's not something we've seen a lot. And of course there's the "goodness only knows what Cromwell chooses to do w/ objects" aspect :)

It looks like the `isObjectLiteral` function in `wdl4s` is used absolutely nowhere  which makes me wonder if there's any support for it at all.